### PR TITLE
Refactor uniswap v2 & implement native related methods

### LINKED ISSUE #13: Uniswap-v2: liquidity addtion with native, swap with native tokens
I.e., porting https://github.com/Uniswap/v2-periphery/blob/69617118cda519dab608898d62aaa79877a61004/contracts/UniswapV2Router02.sol#L77, etc.

### PR DESCRIPTION
**Pull request by HyunggyuJang**
- Resolves #13
- Split wnative contract for native related methods of router
- Factor out library functions from router so that it can be used for other contracts also
- Introduce `ensure!` macro, which has same specification with `require` from solidity
- For test, I'll introduce [typechain-polkadot](https://github.com/Supercolony-net/typechain-polkadot) to cover MVP features for uniswap v2 as a separate PR.

### CODE REVIEW

File: uniswap-v2/logics/impls/router/router.rs [hunk 1]
```diff
@@ -167,185 +221,243 @@ impl<T: Storage<data::Data>> Router for T {
...
+    fn swap_exact_tokens_for_native(
+        &mut self,
+        amount_in: Balance,
+        amount_out_min: Balance,
+        path: Vec<AccountId>,
+        to: AccountId,
+        deadline: u64,
+    ) -> Result<Vec<Balance>, RouterError> {
+        let factory = self.data().factory;
+        let pair_hash_ref = self.data().pair_code_hash;
```
Comment (PierreOssun) [inline uniswap-v2/logics/impls/router/router.rs:347]:
```suggestion
        let pair_hash = self.data().pair_code_hash;
```

File: uniswap-v2/logics/impls/router/router.rs [hunk 2]
```diff
@@ -167,185 +221,243 @@ impl<T: Storage<data::Data>> Router for T {
...
+        let factory_ref = factory.as_ref();
+        let pair_hash_ref = pair_hash_ref.as_ref();
```
Comment (PierreOssun) [inline uniswap-v2/logics/impls/router/router.rs:349]:
```suggestion
        let pair_hash_ref = pair_hash.as_ref();
```

File: uniswap-v2/logics/traits/helper.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+pub fn get_reserves(
+    factory: &[u8; 32],
+    pair_code_hash: &[u8],
+    token_a: AccountId,
+    token_b: AccountId,
+) -> Result<(Balance, Balance), HelperError> {
+    let (token_0, _) = sort_tokens(token_a, token_b)?;
+    // get pair contract address from factory
```
[the remark this answers is not in the document]
Comment (HyunggyuJang) [reply uniswap-v2/logics/traits/helper.rs:71]: Ah, I missed that, thanks.

File: uniswap-v2/logics/traits/helper.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    } else {
+        (token_b, token_a)
+    };
+
+    ensure!(!token_0.is_zero(), HelperError::ZeroAddress);
+
+    Ok((token_0, token_1))
+}
+
+pub fn pair_for(
```
[the remark this answers is not in the document]
Comment (HyunggyuJang) [reply uniswap-v2/logics/traits/helper.rs:44]: Got it, I missed that also.

File: uniswap-v2/logics/traits/wnative.rs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    contracts::psp22::PSP22Error,
+    traits::Balance,
+};
+
+#[openbrush::wrapper]
+pub type WnativeRef = dyn Wnative;
+
+#[openbrush::trait_definition]
+pub trait Wnative {
+    /// @notice Deposit NATIVE to wrap it
```
[the remark this answers is not in the document]
Comment (HyunggyuJang) [reply uniswap-v2/logics/traits/wnative.rs:11]: Yep, I forget to remove it, thanks!

File: uniswap-v2/logics/traits/transfer_helper.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
+use crate::traits::wnative::WnativeRef;
```
[the remark this answers is not in the document]
Comment (HyunggyuJang) [reply uniswap-v2/logics/traits/transfer_helper.rs:1]: Agree. Will reflect it

File: uniswap-v2/logics/impls/router/router.rs [hunk 7]
```diff
@@ -354,134 +466,73 @@ impl<T: Storage<data::Data>> Router for T {
...
-            let amount_a_optimal = self._quote(amount_b_desired, reserve_b, reserve_a)?;
-            if amount_a_desired < amount_a_optimal {
-                return Err(RouterError::ExcessiveAAmount)
-            }
-            if amount_a_optimal < amount_a_min {
-                return Err(RouterError::InsufficientAAmount)
-            }
-
+            let amount_a_optimal = quote(amount_b_desired, reserve_b, reserve_a)?;
+            assert!(amount_a_optimal <= amount_a_desired);
```
[the remark this answers is not in the document]
Comment (HyunggyuJang) [reply uniswap-v2/logics/impls/router/router.rs:492]: Note that this code is only for document purpose as from previous branching, asserted invariant always hold. Thus, I'll replace it with comment.

File: uniswap-v2/logics/traits/transfer_helper.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (HyunggyuJang) [reply uniswap-v2/logics/traits/transfer_helper.rs]: Also moved math to helpers.

File: uniswap-v2/logics/impls/router/router.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (HyunggyuJang) [reply uniswap-v2/logics/impls/router/router.rs]: Done.

File: rust-toolchain.toml [hunk 8]
```diff
@@ -1,5 +1,5 @@
 [toolchain]
-channel = "nightly-2022-08-15"
+channel = "nightly-2022-08-11"
```
Comment (PierreOssun) [inline rust-toolchain.toml:2]: Why this update ?

Comment (HyunggyuJang) [reply rust-toolchain.toml:2]: Ah, it slipped into, I’ll revert it

**Pull request merged**